dll/shellext/zipfldr/CZipExtract.cpp

```diff
@@ -0,0 +1,477 @@
+
+            char CombinedPath[MAX_PATH * 2] = { 0 };
+            PathCombineA(CombinedPath, BaseDirectory, Name);
+            CStringA FullPath = CombinedPath;
+            FullPath.Replace('/', '\\');    /* SHPathPrepareForWriteA does not handle '/' */
+            DWORD dwFlags = SHPPFW_DIRCREATE | (is_dir ? SHPPFW_NONE : SHPPFW_IGNOREFILENAME);
+            HRESULT hr = SHPathPrepareForWriteA(hDlg, NULL, FullPath, dwFlags);
+            if (FAILED_UNEXPECTEDLY(hr))
+            {
+                return false;
+            }
+            CurrentFile++;
+            if (is_dir)
+                continue;
+
+            const char* password = NULL;
+            /* FIXME: Process password, if required and not specified, prompt the user */
+            err = unzOpenCurrentFilePassword(uf, password);
+            if (err != UNZ_OK)
+            {
+                __debugbreak();
+                return false;
+            }
+
+            HANDLE hFile = CreateFileA(FullPath, GENERIC_WRITE, 0, NULL, CREATE_NEW, FILE_ATTRIBUTE_NORMAL, NULL);
+            if (hFile == INVALID_HANDLE_VALUE)
+            {
+                DWORD dwErr = GetLastError();
+                if (dwErr == ERROR_FILE_EXISTS)
+                {
+                    bool bOverwrite = bOverwriteAll;
+                    if (!bOverwriteAll)
+                    {
+                        CConfirmReplace::DialogResult Result = CConfirmReplace::ShowDlg(hDlg, FullPath);
+                        switch (Result)
+                        {
+                        case CConfirmReplace::YesToAll:
+                            bOverwriteAll = true;
+                        case CConfirmReplace::Yes:
+                            bOverwrite = true;
+                            break;
+                        case CConfirmReplace::No:
+                            break;
+                        case CConfirmReplace::Cancel:
+                            return false;
+                        }
+                    }
+
+                    if (bOverwrite)
+                    {
+                        hFile = CreateFileA(FullPath, GENERIC_WRITE, 0, NULL, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, NULL);
+                        if (hFile == INVALID_HANDLE_VALUE)
+                        {
+                            dwErr = GetLastError();
+                        }
+                    }
+                    else
+                    {
+                        unzCloseCurrentFile(uf);
+                        continue;
+                    }
+                }
+                if (hFile == INVALID_HANDLE_VALUE)
+                {
+                    unzCloseCurrentFile(uf);
+                    __debugbreak();
+                    return false;
+                }
+            }
+
+            do
+            {
+                err = unzReadCurrentFile(uf, Buffer, sizeof(Buffer));
+
+                if (err < 0)
+                {
+                    __debugbreak();
```

Comment:
Are these debugbreaks meant to be removed in the near future?

Reply:
yeah, I probably should replace them with DPRINT1 before committing.